### PR TITLE
Add Forge permission handler

### PR DESCRIPTION
**Pull request by kashike**
[SpongeCommon](https://github.com/SpongePowered/SpongeCommon/pull/1461) | **SpongeForge**
#1049

### CODE REVIEW

**Comment by kashike** [comment]
@luacs1998: Could I get your input here? You're familiar with the Forge permission API.

File: src/main/java/org/spongepowered/mod/command/SpongePermissionHandler.java
Comment (yuuka-miya) [inline src/main/java/org/spongepowered/mod/command/SpongePermissionHandler.java]: Does Sponge have the concept of registering permissions?

Still trying to acquaint myself with the Sponge PermissionService, so...

**Comment by yuuka-miya** [comment]
Looks okay to me, but if you want a second opinion - @LatvianModder?

File: src/main/java/org/spongepowered/mod/command/SpongePermissionHandler.java
Comment (kashike) [reply src/main/java/org/spongepowered/mod/command/SpongePermissionHandler.java]: We have the concept of [creating descriptions](https://github.com/SpongePowered/SpongeAPI/blob/bleeding/src/main/java/org/spongepowered/api/service/permission/PermissionService.java#L94-L102) which requires the mod/plugin, the permission node, and a description.

File: src/main/java/org/spongepowered/mod/command/SpongePermissionHandler.java
Comment (kashike) [inline src/main/java/org/spongepowered/mod/command/SpongePermissionHandler.java]: How should we handle this? Sponge has a `Context` concept as well, not sure if we want to convert `IContext` data to a `Context`?

File: src/main/java/org/spongepowered/mod/command/SpongePermissionHandler.java
Comment (kashike) [inline src/main/java/org/spongepowered/mod/command/SpongePermissionHandler.java]: Should we keep track of this, just for Forge mods wishing to get this?

File: src/main/java/org/spongepowered/mod/command/SpongePermissionHandler.java
Comment (ryantheleach) [reply src/main/java/org/spongepowered/mod/command/SpongePermissionHandler.java]: To add on to this, descriptions can designate generic like permissions with placeholders for variable permissions. e.g. nucleus.teleport.world.<dimension> to have per dimension permissions.

**Comment by LatvianModder** [comment]
I guess this works. I wouldnt use that many lambdas in getNodeDescription, but that isn't used frequently anyway, so I guess its fine. Im ok, as long as it properly handles registered permissions. Only not sure why is that try{}catch{} required, just register your handler in PREINIT, and it will be fine, because mods are expected to register nodes in INIT+

File: src/main/java/org/spongepowered/mod/command/SpongePermissionHandler.java
Comment (Zidane) [inline src/main/java/org/spongepowered/mod/command/SpongePermissionHandler.java]: This is bad, if setting the permission handler fails then log it to the console as that likely means someone else is doing so. Let the server admin figure out how to proceed but don't blindly blow away another mod whose purpose is to do permissions.

Comment (LatvianModder) [reply src/main/java/org/spongepowered/mod/command/SpongePermissionHandler.java]: No, I actually agree with Sponge being the strongest permission handler (Config option for this would be nice tho), but I dont agree with aggressively changing the private field. You only should set permission handler in PREINIT state, as I said earlier. Hopefully this is changed

Comment (kashike) [reply src/main/java/org/spongepowered/mod/command/SpongePermissionHandler.java]: @luacs1998 also suggested that we be the strongest. There's the possibility that another mod will replace our handler, so my thought was to do it after pre-init. I can add a config option to *disable* our replacing of other permission handlers if one is registered before Sponge.

File: src/main/java/org/spongepowered/mod/service/permission/SpongePermissionHandler.java
Comment (dualspiral) [inline src/main/java/org/spongepowered/mod/service/permission/SpongePermissionHandler.java]: I put this on a commit that has been blown away rather than the PR, but I'll put it in again - permissions have nothing to do with commands, why is this in a commands package?

File: src/main/java/org/spongepowered/mod/service/permission/SpongePermissionHandler.java
Comment (Xfel) [inline src/main/java/org/spongepowered/mod/service/permission/SpongePermissionHandler.java]: This will only work for online players. Is that intended?

Comment (Zidane) [reply src/main/java/org/spongepowered/mod/service/permission/SpongePermissionHandler.java]: @kashike ?

**Comment by phit** [comment]
@kashike status of this?

**Comment by BrainStone** [comment]
Yes, what’s the status on this? We still need a plugin for this.

**Comment by BrainStone** [comment]
Bump...

**Comment by tonnic** [comment]
please add this....we need this implemented

**Comment by maxanier** [comment]
What exactly has to be tested?
I could try to verify the general functionality by adding some Forge permission checks to a mod of mine. Then build this branch, setup a server with it, my mod and LuckPerms and check if LuckPerms can access the mod's permissions?

**Comment by gabizou** [comment]
> I could try to verify the general functionality by adding some Forge permission checks to a mod of mine. Then build this branch, setup a server with it, my mod and LuckPerms and check if LuckPerms can access the mod's permissions?

Correct

**Comment by maxanier** [comment]
So, I did that and (after I found out I have to enable this in the Sponge configuration) it works fine. I can give/deny the permissions registered by my mod using LuckPerms and they are correctly identified.

Only issue is that, the default permission level passed to the registerNode method is ignored and instead the default permission level is undefined (which results in deny).
Most, if not all, permissions introduced by my mod should be available to all players. So one would have to manually grant "vampirism.*" to every player/group. Before they are actually able to play the mod with this enabled.

However, since this has to be manually enabled anyway, I guess that is ok.
I am very happy there is finally a way to integrate mods with permission plugins.

**Comment by phit** [comment]
you could already do so https://docs.spongepowered.org/stable/en/plugin/permissions.html#for-forge-mods you don't need this PR for that

**Comment by LatvianModder** [comment]
If I had known about this way before the Forge PermissionAPI PR was merged, I would have simply just added a hook in checkPermission() method for forge mods, which would make things 10000x easier. That's why I suggest to try this approach in 1.13/1.14+ and basically just not care about it in 1.12. Mods would most likely not care about it this late anyway, but if new PR was made for 1.13, maybe then things can change

**Comment by BrainStone** [comment]
The Forge Permission API is pretty well used already on 1.12.2. So it would make sense to already implement it in API 7, so people don't have to use an unreleased version of an abandonned plugin that bridges the gap.  
Keep in mind that 1.12.2 will be a long term version for modded MC.

**Comment by BrainStone** [comment]
@gabizou what's the holdup? This PR is a strict improvement of the project, has been confirmed to work and been reviewed by several people.

**Comment by Zidane** [comment]
@phit has objections to merging this and can elaborate why.

**Comment by BrainStone** [comment]
Sure. Let's keep it rotting until it can't be merged anymore.

**Comment by pie-flavor** [comment]
@BrainStone oy. No need to be rude.

**Review by ImMorpheus — APPROVED** [review]
Tested with FTB Utilities.
Tried to change `ftbutilities.chat.nickname.set` with Luckperms, everything is working as intended.

**Pull request merged**
